File: cql3/UserOptions.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.cassandra.cql3;

import java.util.HashMap;
import java.util.Map;

import org.apache.cassandra.auth.IAuthenticator;
import org.apache.cassandra.config.DatabaseDescriptor;
import org.apache.cassandra.exceptions.InvalidRequestException;
import org.apache.cassandra.utils.FBUtilities;

public class UserOptions
{
    private final Map<IAuthenticator.Option, Object> options = new HashMap<IAuthenticator.Option, Object>();

    public void put(String name, Object value)
    {
        options.put(IAuthenticator.Option.valueOf(name.toUpperCase()), value);
    }

    public boolean isEmpty()
    {
        return options.isEmpty();
    }

    public Map<IAuthenticator.Option, Object> getOptions()
    {
        return options;
    }

    public void validate() throws InvalidRequestException
    {
        for (IAuthenticator.Option option : options.keySet())
        {
            if (!DatabaseDescriptor.getAuthenticator().supportedOptions().contains(option))
                throw new InvalidRequestException(String.format("%s doesn't support %s option",
                                                                DatabaseDescriptor.getAuthenticator().getClass().getName(),
                                                                option));
        }
    }

    public String toString()
    {
        return FBUtilities.toString(options);
    }
}

File: cql3/user_options.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/*
 * Modified by ScyllaDB
 *
 * Copyright 2016 ScyllaDB
 */

/*
 * This file is part of Scylla.
 *
 * Scylla is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Scylla is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <string.h>

#include <boost/range/adaptor/map.hpp>

#include "auth/authenticator.hh"
#include "user_options.hh"

void cql3::user_options::put(const sstring& name, const sstring& value) {
    _options[auth::authenticator::string_to_option(name)] = value;
}

void cql3::user_options::validate(const auth::authenticator& a) const {
    for (auto o : _options | boost::adaptors::map_keys) {
        if (!a.supported_options().contains(o)) {
            throw exceptions::invalid_request_exception(
                            sprint("%s doesn't support %s option",
                                            a.qualified_java_name(),
                                            a.option_to_string(o)));
        }
    }
}

